lib/node_modules/@stdlib/fs/resolve-parent-paths/lib/async.js

```diff
@@ -0,0 +1,425 @@
+*/
+
+'use strict';
+
+// MODULES //
+
+var resolve = require( 'path' ).resolve;
+var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
+var isFunction = require( '@stdlib/assert/is-function' );
+var cwd = require( '@stdlib/process/cwd' );
+var exists = require( '@stdlib/fs/exists' );
+var filled = require( '@stdlib/array/base/filled' );
+var format = require( '@stdlib/string/format' );
+var validate = require( './validate.js' );
+
+
+// VARIABLES //
+
+var MODES = {
+	'first': first,
+	'some': some,
+	'all': all,
+	'each': each
+};
+
+
+// FUNCTIONS //
+
+/**
+* Asynchronously resolves the first path match from a set of paths by walking parent directories.
+*
+* @private
+* @param {Array<string>} paths - paths to resolve
+* @param {string} dir - base directory
+* @param {Callback} done - callback to invoke after resolving paths
+* @returns {void}
+*/
+function first( paths, dir, done ) {
+	var child;
+	var spath;
+	var FLG = 0; // flag to track if we are done traversing a directory level
+	var out;
+	var i;
+
+	// Start at a base directory and continue moving up through each parent directory...
+	out = [];
+	for ( i = 0; i < paths.length; i++ ) {
+		spath = resolve( dir, paths[ i ] );
+		exists( spath, getCallback( spath ) );
+	}
+
+	/**
+	* Returns a callback to be invoked upon checking for path existence.
+	*
+	* @private
+	* @param {string} spath - resolved path
+	* @returns {Callback} callback
+	*/
+	function getCallback( spath ) {
+		return onExists;
+
+		/**
+		* Callback invoked after checking for path existence.
+		*
+		* @private
+		* @param {(Error|null)} error - error object
+		* @param {boolean} bool - boolean indicating if a path exists
+		* @returns {void}
+		*/
+		function onExists( error, bool ) { // eslint-disable-line node/handle-callback-err
+			var j;
+			if ( bool ) {
+				out.push( spath );
+			}
+			// If we have traversed all paths at the current directory level, resolve parent directory...
+			FLG += 1;
+			if ( FLG === paths.length ) {
+				// Check if we have resolved any path...
+				if ( out.length > 0 ) {
+					return done( null, [ out[ 0 ] ] );
+				}
+				// Resolve a parent directory:
+				child = dir;
+				dir = resolve( dir, '..' );
+
+				// Reset flag:
+				FLG = 0;
+
+				// If we have already reached root, we cannot resolve any higher directories...
+				if ( child === dir ) {
+					return done( null, out );
+				}
+				// Resolve paths at next directory level:
+				for ( j = 0; j < paths.length; j++ ) {
+					spath = resolve( dir, paths[ j ] );
+					exists( spath, getCallback( spath ) );
+				}
+			}
+		}
+	}
+}
+
+/**
+* Asynchronously resolves one or more paths from a set of paths at a directory level by walking parent directories.
+*
+* @private
+* @param {Array<string>} paths - paths to resolve
+* @param {string} dir - base directory
+* @param {Callback} done - callback to invoke after resolving paths
+* @returns {void}
+*/
+function some( paths, dir, done ) {
```

Comment:
One thing, this can output paths in a different order than in the original array with the current implementation due to it's asynchronous nature. That should be fine, as it's not expected to output elements corresponding to the input paths..?